homeassistant/components/sensor/gitlab_ci.py

```diff
@@ -0,0 +1,187 @@
+
+    def update(self):
+        """Collect updated data from GitLab API."""
+        self._gitlab_data.update()
+
+        self._status = self._gitlab_data.status
+        self._started_at = self._gitlab_data.started_at
+        self._finished_at = self._gitlab_data.finished_at
+        self._duration = self._gitlab_data.duration
+        self._commit_id = self._gitlab_data.commit_id
+        self._commit_date = self._gitlab_data.commit_date
+        self._build_id = self._gitlab_data.build_id
+        self._branch = self._gitlab_data.branch
+        self._state = self._gitlab_data.status
+        self._available = self._gitlab_data.available
+
+
+class GitLabData():
+    """GitLab Data object."""
+
+    def __init__(self, gitlab_id, priv_token, interval, url):
+        """Fetch data from GitLab API for most recent CI job."""
+        self._gitlab_id = gitlab_id
+        self._priv_token = priv_token
+        self._interval = interval
+        self._url = url
+        self.update = Throttle(interval)(self._update)
+
+        self._response = None
+        self._response_json = None
+        self.available = False
+        self.status = STATE_UNKNOWN
+        self.started_at = STATE_UNKNOWN
+        self.finished_at = STATE_UNKNOWN
+        self.duration = STATE_UNKNOWN
+        self.commit_id = STATE_UNKNOWN
+        self.commit_date = STATE_UNKNOWN
+        self.build_id = STATE_UNKNOWN
+        self.branch = STATE_UNKNOWN
+        self.state = STATE_UNKNOWN
+
+    def _update(self):
+        _logger = logging.getLogger(__name__)
+        _logger.debug(self._interval)
+        import gitlab
+        import requests
+        try:
+            _gitlab = gitlab.Gitlab(
+                self._url, private_token=self._priv_token)
+            _gitlab.auth()
+            _projects = _gitlab.projects.get(self._gitlab_id)
+            _last_pipeline = _projects.pipelines.list()[0]
+            _last_job = _last_pipeline.jobs.list()[0]
+            self.status = _last_pipeline.attributes.get('status')
+            self.started_at = _last_job.attributes.get('started_at')
+            self.finished_at = _last_job.attributes.get('finished_at')
+            self.duration = _last_job.attributes.get('duration')
+            _commit = _last_job.attributes.get('commit')
+            self.commit_id = _commit.get('id')
+            self.commit_date = _commit.get('committed_date')
+            self.build_id = _last_job.attributes.get('id')
+            self.branch = _last_job.attributes.get('ref')
+            self.state = self.status
+            self.available = True
+        except gitlab.exceptions.GitlabAuthenticationError as erra:
+            _logger.error("Authentication Error: %s", erra)
+            self.available = False
+        except gitlab.exceptions.GitlabGetError as errg:
+            _logger.error("Project Not Found: %s", errg)
+            self.available = False
+        except requests.exceptions.ConnectionError as errc:
```

Comment:
Would this not be covered by the Gitlab exceptions?

Reply:
Removing, this catch is never actually encountered as the component fails to even be set up if an invalid URL is provided.